### ISSUE TITLE
Performance issue: AutoScheduler (with CURLMOPT_MAX_TOTAL_CONNECTIONS) is slower than ManualScheduler

### ISSUE THREAD

**Issue by mpyw** [mpyw/co#24]
[AutoScheduler](https://github.com/mpyw/co/blob/master/src/Internal/AutoScheduler.php) (with `CURLMOPT_MAX_TOTAL_CONNECTIONS`) may have two major issues.

- **<del>It is slower than [ManualScheduler](https://github.com/mpyw/co/blob/master/src/Internal/ManualScheduler.php)</del>**.
- It cannot properly calculate times related to `CURLINFO_*`.

There have been no reliable benchmarks  yet, so we'd better measure them.

**Comment by mpyw** [comment]
I had some experiment on this issue.

# Executor Sources

## C

Compilation command: `gcc request.c -lcurl -o request` 

```c
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#ifndef WIN32
#include <unistd.h>
#endif
#include <curl/multi.h>
 
static const char *urls[] = {
  "http://www.microsoft.com",
  "http://www.opensource.org",
  "http://www.google.com",
  "http://www.yahoo.com",
  "http://www.ibm.com",
  "http://www.mysql.com",
  "http://www.oracle.com",
  "http://www.ripe.net",
  "http://www.iana.org",
  "http://www.amazon.com",
  "http://www.netcraft.com",
  "http://www.cnet.com",
  "http://www.news.com",
  "http://www.cnn.com",
  "http://www.dell.com",
  "http://www.hp.com",
  "http://www.mit.edu",
  "http://www.nist.gov",
  "http://www.ebay.com",
  "http://www.playstation.com",
  "http://www.ieee.org",
  "http://www.apple.com",
  "http://www.symantec.com",
  "http://www.zdnet.com",
  "http://www.fujitsu.com",
  "http://www.hotmail.com",
  "http://www.ecma.com",
  "http://www.msn.com",
};
 
#define CNT sizeof(urls)/sizeof(char*)
 
static size_t cb(char *d, size_t n, size_t l, void *p)
{
  (void)d;
  (void)p;
  return n*l;
}
 
static void init(CURLM *cm, int i)
{
  CURL *eh = curl_easy_init();
  curl_easy_setopt(eh, CURLOPT_WRITEFUNCTION, cb);
  curl_easy_setopt(eh, CURLOPT_URL, urls[i]);
  curl_easy_setopt(eh, CURLOPT_PRIVATE, urls[i]);
  curl_multi_add_handle(cm, eh);
}
 
int main(int argc, const char *argv[])
{
  CURLM *cm;
  CURLMsg *msg;
  long L;
  unsigned int C=0;
  int M, Q, U = -1;
  fd_set R, W, E;
  struct timeval T;
  int use_total = argc > 1 ? atoi(argv[1]) : 0;
  int max = argc > 2 ? atoi(argv[2]) : 0;
 
  curl_global_init(CURL_GLOBAL_ALL);
 
  cm = curl_multi_init();
  
  if (use_total) {
    curl_multi_setopt(cm, CURLMOPT_MAX_TOTAL_CONNECTIONS, max);
    for(C = 0; C < CNT; ++C) {
      init(cm, C);
    }
  } else {
    for(C = 0; C < max; ++C) {
      init(cm, C);
    }
  }
 
  while(U) {
    curl_multi_perform(cm, &U);
 
    if(U) {
      FD_ZERO(&R);
      FD_ZERO(&W);
      FD_ZERO(&E);
 
      if(curl_multi_fdset(cm, &R, &W, &E, &M)) {
        fprintf(stderr, "E: curl_multi_fdset\n");
        return EXIT_FAILURE;
      }
 
      if(curl_multi_timeout(cm, &L)) {
        fprintf(stderr, "E: curl_multi_timeout\n");
        return EXIT_FAILURE;
      }
      if(L == -1)
        L = 100;
 
      if(M == -1) {
#ifdef WIN32
        Sleep(L);
#else
        sleep((unsigned int)L / 1000);
#endif
      }
      else {
        T.tv_sec = L/1000;
        T.tv_usec = (L%1000)*1000;
 
        if(0 > select(M+1, &R, &W, &E, &T)) {
          fprintf(stderr, "E: select(%i,,,,%li): %i: %s\n",
              M+1, L, errno, strerror(errno));
          return EXIT_FAILURE;
        }
      }
    }
 
    while((msg = curl_multi_info_read(cm, &Q))) {
      if(msg->msg == CURLMSG_DONE) {
        char *url;
        CURL *e = msg->easy_handle;
        curl_easy_getinfo(msg->easy_handle, CURLINFO_PRIVATE, &url);
        fprintf(stderr, "R: %d - %s <%s>\n",
                msg->data.result, curl_easy_strerror(msg->data.result), url);
        curl_multi_remove_handle(cm, e);
        curl_easy_cleanup(e);
      }
      else {
        fprintf(stderr, "E: CURLMsg (%d)\n", msg->msg);
      }
      if(C < CNT) {
        init(cm, C++);
        U++;
      }
    }
  }
 
  curl_multi_cleanup(cm);
  curl_global_cleanup();
 
  return EXIT_SUCCESS;
}
```

## PHP 

```php
#!/usr/bin/env php
<?php

$urls = [
  "http://www.microsoft.com",
  "http://www.opensource.org",
  "http://www.google.com",
  "http://www.yahoo.com",
  "http://www.ibm.com",
  "http://www.mysql.com",
  "http://www.oracle.com",
  "http://www.ripe.net",
  "http://www.iana.org",
  "http://www.amazon.com",
  "http://www.netcraft.com",
  "http://www.cnet.com",
  "http://www.news.com",
  "http://www.cnn.com",
  "http://www.dell.com",
  "http://www.hp.com",
  "http://www.mit.edu",
  "http://www.nist.gov",
  "http://www.ebay.com",
  "http://www.playstation.com",
  "http://www.ieee.org",
  "http://www.apple.com",
  "http://www.symantec.com",
  "http://www.zdnet.com",
  "http://www.fujitsu.com",
  "http://www.hotmail.com",
  "http://www.ecma.com",
  "http://www.msn.com",
];
 
$CNT = count($urls);
  
function init($cm, $i)
{
  $ch = curl_init();
  curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
  curl_setopt($ch, CURLOPT_URL, $GLOBALS['urls'][$i]);
  curl_setopt($ch, CURLOPT_PRIVATE, $GLOBALS['urls'][$i]);
  curl_multi_add_handle($cm, $ch);
}
 
  $C=0;
  $U = -1;
  $use_total = $argc > 1 ? intval($argv[1]) : 0;
  $max = $argc > 2 ? intval($argv[2]) : 0;
 
  $cm = curl_multi_init();
  
  if ($use_total) {
    curl_multi_setopt($cm, CURLMOPT_MAX_TOTAL_CONNECTIONS, $max);
    for($C = 0; $C < $CNT; ++$C) {
      init($cm, $C);
    }
  } else {
    for($C = 0; $C < $max; ++$C) {
      init($cm, $C);
    }
  }
 
  while($U) {
    curl_multi_exec($cm, $U);
 
    if($U) {
      if(curl_multi_select($cm, 0.002) === -1) {
        usleep(2500);
        continue;
      }
    }
 
    while($msg = curl_multi_info_read($cm)) {
      $url = curl_getinfo($msg['handle'], CURLINFO_PRIVATE);
      fprintf(STDERR, "R: %d - %s <%s>\n",
              $msg['result'], curl_error($msg['handle']) ?: 'No Errors', $url);
      curl_multi_remove_handle($cm, $msg['handle']);
      curl_close($msg['handle']);
      if($C < $CNT) {
        init($cm, $C++);
        $U++;
      }
    }
  }
 
  curl_multi_close($cm);
```

# Benchmarker Source

## Python 3

```python
from matplotlib import pyplot as plt
from time import time
import subprocess, os, sys

def eprint(*args, **kwargs):
    print(*args, file=sys.stderr, **kwargs)

def calc_time(args):
    start = time()
    subprocess.Popen(args).communicate()
    end = time()
    return end - start

def calc_average(auto, limit):
    s = 0.0
    for i in range(5):
        eprint(
            "\n\n\nProcessing auto:{}, limit:{}, round:{}/{}\n"
            .format(auto, limit, i + 1, 5)
        )
        # CHANGE THIS LINE
        s += calc_time(['./request OR ./request.php', '1' if auto else '0', str(limit)])
    return s / 5

def calc_series(auto):
    x = []
    y = []
    for limit in range(28, 0, -1):
        x.append(limit)
        y.append(calc_average(auto, limit))
    return x, y

if __name__ == '__main__':
    colors = ['b', 'g']
    settings = [
        {
            'name': 'Auto scheduling with CURLMOPT_MAX_TOTAL_CONNECTIONS',
            'args': [True],
        },
        {
            'name': 'Manual scheduling',
            'args': [False],
        },
    ]
    for i, setting in enumerate(settings):
        x, y = calc_series(*setting['args'])
        plt.plot(
            x, y, '-',
            color=colors[i % len(colors)],
            label=setting['name']
        )
    plt.title('CURLM scheduling performance benchmarks [PHP]')
    plt.xlabel('Max TCP connections')
    plt.ylabel('Seconds')
    plt.legend(loc='best')
    plt.savefig('result-php.eps')
    plt.clf()
```

# Results

(Sorry, x-axis should be opposite direction)

## C

![result](https://cloud.githubusercontent.com/assets/1351893/18007325/8715e4ec-6bde-11e6-95d7-d02529742089.png)

## PHP

![result-php](https://cloud.githubusercontent.com/assets/1351893/18007328/8b07f342-6bde-11e6-8d87-3b2b98a42fdc.png)

# Conclusion

**There are no remarkable differences on the performance.** In fact, however, `CURLINFO_TOTAL_TIME` is could not be properly calculated. So we need option to choose auto or manual scheduling.

**Comment by mpyw** [comment]
Implemented on b4df70c
